### ISSUE TITLE
Blob constructor is not spec compliant

### ISSUE THREAD

**Issue by shamilovtim** [facebook/react-native#44125]
### Description

The constructor for React Native's [Blob has a comment](https://github.com/facebook/react-native/blob/8c53ac60782d8ccd2ccd7706d41e60cda93e2988/packages/react-native/Libraries/Blob/Blob.js#L57):
> Currently we only support creating Blobs from other Blobs.

BlobManager [confirms this:](https://github.com/facebook/react-native/blob/main/packages/react-native/Libraries/Blob/BlobManager.js#L71)
"Creating blobs from 'ArrayBuffer' and 'ArrayBufferView' are not supported"

This API is not spec compliant, since this is what the Blob specification [specifies](https://developer.mozilla.org/en-US/docs/Web/API/Blob/Blob): 

> (Constructed from) An [iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterable_protocol) object such as an [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array), having [ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer)s, [TypedArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray)s, [DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView)s, [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob)s, strings, or a mix of any of such elements, that will be put inside the [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob). Strings should be well-formed Unicode, and lone surrogates are sanitized using the same algorithm as [String.prototype.toWellFormed()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/toWellFormed).

### Steps to reproduce

1. Try to call:

```ts
new Blob(new Uint8Array([
    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20,
    21, 22, 23,
  ]);
  ```
  
  See that RN doesn't support Blobs composed of anything but Blobs or strings.

### React Native Version

0.73.6

### Affected Platforms

Runtime - Android, Runtime - iOS

### Output of `npx react-native info`

```text
N/A
```

### Stacktrace or Logs

```text
N/A
```

### Reproducer

N/A

**Comment by huzhanbo1996** [comment]
Hi @shamilovtim, I'm looking into this. Blob on RN is String based and does not provide any stream/binary API. Could you describe your usage?

**Comment by shamilovtim** [comment]
Hi @huzhanbo1996 

The problem is that everyone on web builds on the Blob specification which is available on other platforms and browsers but RN's Blob does not meet these specifications. Then when libraries are used in RN they break because they rely on broken / incomplete Blob.

To describe what I need

1. I need to be able to construct a Blob from TypedArrays or ArrayBuffers like the Blob specification allows for. Because this is missing from RN Blob, certain web libraries break when used on react native.

> Blob on RN is String based and does not provide any stream/binary API
2. Yes this is also a separate problem. I need to be able to call `stream()` on Blob as well, and this is another case where RN Blob does not match web specs.

**Comment by LinusU** [comment]
Somewhat related to this, and specifically related to this part:

> Blob on RN is String based and does not provide any stream/binary API.

I guess that this is the reason that [`Response#arrayBuffer()`](https://developer.mozilla.org/en-US/docs/Web/API/Response/arrayBuffer) doesn't work, which in turn makes [`ky`](https://github.com/sindresorhus/ky) not work (https://github.com/sindresorhus/ky/issues/566).

**Comment by shamilovtim** [comment]
Just to forestall any duplicate efforts to fix I think @grabbou is working on a real Blob implementation

**Comment by grabbou** [comment]
Yeah, thanks for mentioning me. I am indeed working (releasing this week - I hope) a spec complaint Blob with both steam and array buffer and a bunch of other improvements.

Please reach out to me grabbou@gmail.com and we will coordinate to make sure it does fix your issues.

It's not something I'll be sending as a PR to RN Core since it's based on Nitro. 

However, as a short term hot fix, I think you can polyfill missing methods. In case of arrayBuffer, just create one by using text encoder and string value of Blob.

**Comment by kirill-dev-pro** [comment]
Hey @grabbou, how's the progress with this issue?

**Comment by grabbou** [comment]
We're working on support for ArrayBuffers in Turbo Modules, Blob will soon be spec compliant (first, we land support in TM, then follow-up work on modules). 

Hard to say what is the ETA, but I hope this quarter it happens!
